docs/fendermint/general_cross_messages.md

```diff
@@ -0,0 +1,120 @@
+## Overview
+
+General cross messages enable the transmission of arbitrary messages between IPC subnets. They are particularly useful for transferring tokens (when supported, more on this later) or transmitting arbitrary data. This is not a feature that allows users to trigger messages from wallets; rather, it is a framework for cross-subnet contract-to-contract communication.
+
+Messages are transmitted across subnets using top-down and bottom-up mechanisms. For example, if a message is sent from a parent to a child subnet, the top-down mechanism is used, and vice versa. Similarly, messages between subnets not directly connected but sharing a common parent are routed automatically by the IPC system.
+
+## Limitations
+
+- Only contracts (not externally owned accounts, or EOAs) can invoke cross messages.
+- Token transfers are only allowed if the source subnet, destination subnet, and all intermediate subnets share the same token.
+- Since cross messages flow through the entire route between the source and destination subnets, multi-hop messages may take longer to be delivered.
+
+## Usage
+
+Messages are passed via the `Gateway` contract, which exposes the following method for triggering cross messages (only callable by another contract):
+
+```solidity
+function sendContractXnetMessage(
+    IpcEnvelope calldata envelope
+) external payable returns (IpcEnvelope memory committed);
+```
+
+The actual message is embedded within the IPC envelope as ABI-encoded bytes.
+
+### IPC Envelope Structure
+
+```solidity
+struct IpcEnvelope {
+    /// @dev Type of the message being propagated.
+    IpcMsgKind kind;
+    /// @dev Destination of the message.
+    IPCAddress to;
+    /// @dev Address of the sender.
+    IPCAddress from;
+    /// @dev Outgoing nonce for the envelope, ensuring uniqueness per subnet and preventing replay attacks.
+    uint64 localNonce;
+    /// @dev Value being sent with the message (e.g., tokens).
+    uint256 value;
+    /// @dev ABI-encoded message payload.
+    bytes message;
+    /// @dev Original nonce from the source subnet, useful for tracing messages across subnets.
+    uint64 originalNonce;
+}
+```
+
+### Message Types of Interest
+
+- **Call**: Represents a request sent to a destination contract.
+- **Result**: Represents the response to a `Call` message, which could either be the result or an error report.
+
+### Contract Requirements
+
+Both sending and receiving contracts must implement the `IpcExchange` abstract contract from the IPC SDK to handle cross messages. Key methods to implement include:
```

Comment:
The correct term is "cross-net messages" or "xnet messages". "Cross messages" was a poor past usage we're trying to deprecate.